distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p/>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p/>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.test.itests.common.downloads;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Special async stream reader for synchronizing with other async readers of the same type.
+ * Will block-wait for the barrier to fire on the specified index until it continues processing.
+ */
+public class BlockingInputStreamReader extends AbstractIndexBlocker implements AutoCloseable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BlockingInputStreamReader.class);
+
+    private final AtomicInteger currentByteConsumptionIndex = new AtomicInteger(0);
+
+    private final StringBuilder data = new StringBuilder();
+
+    private InputStream inputStream;
+
+    /**
+     * Default constructor.
+     *
+     * @param inputStream The stream to read from.
+     * @param barrier     The barrier to synchronize against.
+     */
+    public BlockingInputStreamReader(String metacardId, InputStream inputStream,
+            CyclicBarrier barrier) {
+        super(metacardId, barrier);
+        this.inputStream = inputStream;
+    }
+
+    @Override
+    public void close() throws IOException {
+        inputStream.close();
```

Comment:
It is rare, though not necessarily wrong, for the consumer of an InputStream to be responsible for closing it. Usually it is the responsibility of the originator of the stream to be responsible for its lifecycle.